feature/177 엔티티 삭제 시, DB에서 "삭제" 상태가 되도록 변경한다

yebink opened the pull request:
Resolves #177

- Post, PostLike, Comment, CommentLike 엔티티에 EntityStatus 필드를 추가했습니다.
- 관련된 깨지는 테스트들도 모두 수정했습니다.
- 중간중간 보이는 불필요한 메소드, 잘못된 로직도 리팩토링했으니 확인 부탁해요!

이미 배포된 엔티티에 status를 추가하는 문제는 직접 컬럼 추가 쿼리를 치는 것 외에는 떠오르지 않네요 🤔
혹시 방법 있다면 공유 부탁합니다~

방학이니 리뷰는 천천히 해주세요 🙂 즐거운 방학 보내세요!

yebink committed:
feature/177 feat: EntityStatus 클래스 생성

yebink committed:
feature/177 feat: Post, Comment, PostLike 엔티티에 EntityStatus 추가

 - PostService.deletePost 로직 변경
 - PostRepository의 불필요한 메소드 제거
 - 기존 delete 관련 깨지는 테스트 수정

yebink committed:
feature/177 feat: PostLikeService.deletePostLike 로직 변경

 - PostRepository에 불필요한 메소드 제거
 - 깨지는 테스트 수정

yebink committed:
feature/177 feat: CommentLike 엔티티에 status 추가

 - Comment.delete에서도 CommentLike 함께 지우도록 변경
 - CommentService.deleteComment 로직 변경
 - CommentLikeService.deleteCommentLike 로직 변경
 - 관련 깨지는 테스트 수정

yebink committed:
feature/177 fix: 기존 Comment.setAuthor 로직 수정

 - null 체크 대상을 author -> this.author로 변경

hwookim commented on src/main/java/com/grasshouse/dorandoran/post/domain/PostLike.java:
```diff
@@ -45,6 +51,10 @@ private void setPost(Post post) {
         }
     }
 
+    public void delete() {
+        this.status = EntityStatus.DELETED;
+    }
+
```
postLike의 경우에는 지웠다가 다시 누를 수 있을텐데 delete 상태 관리가 효율적일까요?
지금은 DELETE 상태가 되어도 POST에 객체가 남아있고 프론트에서는 그걸 모르니 삭제가 이뤄지질 않네요.

YebinK replied:
동의합니다. 더군다나 postlike, commentlike 생성/삭제 api가  toggle api로 변경된다면요.
반영하겠습니다 🙂

hwookim reviewed (changes requested):
상태가 DELETE로 변한다면 해당 객체는 클라이언트한테 넘긴 뒤 클라이언트가 필터링 해야 할까요?
아니면 서버가 DELETE 상태의 객체는 필터링한 후 클라이언트에게 살아있는 객체만 넘겨야 할까요?

아직 해당 PR은 이 부분이 해결되지 않아 모든 엔티티의 삭제 로직이 작동하지 않게 되었네요.😭

리뷰로도 남겼듯 좋아요에 대해서는 적용 여부를 고민해봐야 할 것 같아요.
#177 이슈의 목적에 좋아요도 부합할까요? 🤔

hwookim commented on src/main/java/com/grasshouse/dorandoran/comment/domain/Comment.java:
```diff
@@ -59,6 +62,9 @@
     @OneToMany(mappedBy = "comment", cascade = CascadeType.ALL, orphanRemoval = true)
     private Set<CommentLike> likes = new HashSet<>();
 
+    @Enumerated(value = EnumType.STRING)
+    private EntityStatus status = EntityStatus.ALIVE;
```
이부분을 놓쳤었네요!
@Builder.Default가 필요하지 않을까요?

---
약간 이번 이슈와는 동떨어졌긴 한데... Builder가 클래스에도 붙어있고 생성자에도 붙어있네요.
한쪽에만 있어도 되지 않나요?

yebink committed:
feature/177 refactor: PostLike, CommentLike에서 EntityStatus 제거

LTTTTTE reviewed (commented):
아직 댓글은 "삭제" 상태가 아닌 삭제가 되는거 같군요.
글과 댓글은 초기 status값이 null 입니다.

yebink committed:
feature/177 fix: Post, Comment의 EntityStatus 기본값 설정

 - @Builder.Default 추가

yebink committed:
feature/177 fix: 삭제된 Post를 필터링하는 로직 추가

yebink committed:
feature/177 refactor: PostLike, CommentLike 삭제 로직 변경

 - Post를 삭제하면 -> Post와 Comment들도 삭제 상태로 변경되고, PostLike와 CommentLike들은 삭제된다.
 - Comment를 삭제하면 -> Comment가 삭제 상태로 변경되고, CommentLike들은 삭제된다.

yebink committed:
feature/177 feat: Post의 alive 상태의 comments를 가져오는 querydsl 메소드 작성

yebink committed:
feature/177 refactor: PostRepositorySupport.findPostById 로직 수정

 - test 시에도 쿼리 볼 수 있도록 변경

yebink committed:
feature/177 feat: PostResponse, CommentResponse에 EntityStatus 추가

yebink committed:
feature/177 refactor: QueryDSL 메소드에 상태 필터링 로직 추가

yebink committed:
feature/177 fix: PostRepositorySupport.findPostWithKeywordAndDate에 distinct 추가

hwookim commented on src/main/java/com/grasshouse/dorandoran/comment/service/CommentService.java:
```diff
@@ -47,12 +47,12 @@ private Comment convertToComment(CommentCreateRequest request, Member member) {
         Post post = comment.getPost();
         post.removeComment(comment);
+        comment.delete();
+        System.err.println("====transactinal끝, post delete시작");
```
디버그용으로 만드셨던 코드는 제거해주세요~😃

YebinK replied:
아이코! 😅

hwookim commented on src/main/java/com/grasshouse/dorandoran/post/repository/PostRepositorySupport.java:
```diff
@@ -15,36 +18,69 @@
             .fetch();
     }
 
+    public Post findPostContainingComments(Long postId) {
+        return jpaQueryFactory.selectFrom(post)
+            .innerJoin(post.comments)
+            .fetchJoin()
+            .where(post.id.eq(postId))
+            .fetchFirst();
+    }
+
+    public Post findPostContainingLikes(Long postId) {
+        return jpaQueryFactory.selectFrom(post)
+            .leftJoin(post.likes)
+            .fetchJoin()
+            .where(post.id.eq(postId))
+            .fetchFirst();
+    }
```
어떤 경우에 `innerJoin`이고 어떤 경우에는 `leftJoin`인 건가요??😢
같은 경우라면 하나로 통일하는 게 맞지 않을까요?

쿼리문도 참 어렵네요....

YebinK replied:
post와 comment를 `innerJoin`할 경우, 
post.comments가 비어있다면 post는 결과에 포함되지 않습니다.
`leftJoin` (= `leftOuterJoin`)은 post.comments가 비어있어도 post를 가져오는 방식인데요!

현재로서는 통일해도 무방하지만, 메소드 이름만 본다면 지금 상태가 더 적합하겠는데요!
findPostContainingLikes는 말그대로 likes를 포함하고 있는 post를 가져와야 하기 때문입니다. 🙂 
항상 구분해서 써야 하기에 그대로 두면 더욱 공부가 될 수도 있을 것 같은데요!

이 부분은 팀원들과 함께 이야기해봅시다~

hwookim commented on src/main/java/com/grasshouse/dorandoran/post/service/PostService.java:
```diff
@@ -49,23 +49,18 @@ public PostResponse showPost(Long id) {
-        return postRepository.findById(id)
-            .orElseThrow(PostNotFoundException::new);
+        post.delete();
     }
```
이렇게 보니 `postRepositorySupport`를 사용한 `findById`에서는 예외처리가 안되고 있네요.
만둘어둔 예외를 사용하기 위해서는 `Optional<Post>`를 반환하도록 변경해야하겠어요.

YebinK replied:
동의합니다.
다만, Optional을 반환하게 하면 이 메소드를 사용하는 모든 곳에서 예외 처리를 해주어야 하는데요,
그리고 그 예외 처리에 분기가 들어갈 것 같지 않네요 🤔  (모두 PostNotFoundException이겠지요?)

그래서 메소드 내에서 예외처리하는 방식으로 변경했습니다!

hwookim commented on src/main/java/com/grasshouse/dorandoran/post/repository/PostRepositorySupport.java:
```diff
@@ -15,36 +18,69 @@
-        LocalDateTime endDate) {
-        return jpaQueryFactory
-            .selectFrom(post)
+    public List<Post> findPostWithKeywordAndDate(String keyword, LocalDateTime startDate, LocalDateTime endDate) {
```
```suggestion
    public List<Post> findPostsWithKeywordAndDate(String keyword, LocalDateTime startDate, LocalDateTime endDate) {
```

List니 Posts로 바꾸는 건 어떨까요?

hwookim commented on src/main/java/com/grasshouse/dorandoran/post/repository/PostRepositorySupport.java:
```diff
@@ -15,36 +18,69 @@
 
-    public List<Post> findPostsInBounds(Double leftBound, Double rightBound, Double lowerBound,
-        Double upperBound) {
+    public List<Post> findPostsInBounds(Double leftBound, Double rightBound, Double lowerBound, Double upperBound) {
```
사용되지 않는 메소드를 미리미리 안지워서 부가적인 테스크가 생겼었네요....😢
죄송합니다ㅠㅠ

YebinK replied:
ㅋㅋ괜찮습니다.
지우는거 짜릿하네요

hwookim commented on src/main/java/com/grasshouse/dorandoran/post/repository/PostRepositorySupport.java:
```diff
@@ -15,36 +18,69 @@
+            .fetchFirst();
+    }
+
+    private BooleanExpression postIsAlive() {
+        return post.status.eq(EntityStatus.ALIVE);
+    }
+
+    private BooleanExpression commentIsAlive() {
+        return comment.status.eq(EntityStatus.ALIVE);
+    }
```
`isPost(Comment)Alive` 가 더 맞지 않을까요?

hwookim reviewed (changes requested):
밤 늦게까지 정말 고생 많으셨어요...!😭
마지막 커밋이 오전 4시 45분이라니..... 쉬고 있는 거 맞죠??

몇가지 리뷰 달았으니 확인 부탁드려요!
진짜진짜 고생 많으셨습니다 ㅠㅠ

yebink committed:
feature/177 refactor: 사용되지 않는 api 및 메소드 제거

 - (백) PostController.showPostsInBounds 제거
 - (백) PostService.showPostsInBounds 제거
 - (백) PostRepositorySupport.findPostsInBounds 제거
 - (백) 제거된 메소드들의 테스트도 제거
 - (프론트) api.loadPostsInBounds 제거

chws commented on src/main/java/com/grasshouse/dorandoran/comment/domain/Comment.java:
```diff
@@ -82,7 +89,16 @@ private void setPost(Post post) {
         return author.isSameMember(member);
     }
+
+    public boolean isAlive() {
```
이 메소드는 쓰이지 않는 것 같아요!

chws commented on src/test/java/com/grasshouse/dorandoran/post/repository/PostRepositorySupportTest.java:
```diff
@@ -0,0 +1,107 @@
+
+        Post post = SAVE_POST(member);
+
+        //댓글 2개를 추가한다.
```
불필요한 주석은 제거해주세요:)

YebinK replied:
테스트 메소드 로직이 길어져서 가독성을 위해 넣어둔 것인데... 🤔 
다함께 이야기해봅시다!

YebinK replied:
이야기 결과 제거하기로 했습니다 ~

yebink committed:
feature/177 fix: 삭제된 댓글만 포함된 Post의 경우 null 반환 문제 해결

 - comment.delete 로직을 post.removeComment로 이동

chws reviewed (approved):
엘리, 정말 많은 작업을 하셨네요! 특히 querydsl 부분은👍
다른 팀원들이 리뷰를 많이 해줘서 코드가 한결 나아진 것 같아요:)
몇가지만 코멘트 달았어요~ 수고하셨습니다.

yebink committed:
feature/177 refactor: 조인 전략 통일 및 메소드명 수정

 - PostRepositorySupport.findPostContainingComments & PostRepositorySupport.findPostContainingLikes join 전략을 innerJoin으로 통일
 - postIsAlive() -> isPostAlive()

yebink committed:
feature/177 refactor: 테스트 메소드에 주석 제거

hwookim commented on src/main/java/com/grasshouse/dorandoran/post/controller/PostFilterController.java:
```diff
@@ -18,9 +18,8 @@
-    public ResponseEntity<List<PostResponse>> showFilteredResults(
-        @PostFilter PostFilterRequest request) {
-        List<PostResponse> responses = PostFilterService.showSearchResults(request);
+    public ResponseEntity<List<PostResponse>> showFilteredResults(@PostFilter PostFilterRequest request) {
+        List<PostResponse> responses = PostFilterService.showFilteredResults(request);
```
얘는 왜 바뀐건가욤?

YebinK replied:
바뀐 code convention을 적용했습니다 🙂

hwookim commented on src/main/java/com/grasshouse/dorandoran/post/service/PostFilterService.java:
```diff
@@ -17,9 +17,8 @@
-    public List<PostResponse> showSearchResults(PostFilterRequest request) {
-        List<Post> searchResults = postRepositorySupport
-            .findPostWithKeywordAndDate(request.getKeyword(), request.getStartDate(),
+    public List<PostResponse> showFilteredResults(PostFilterRequest request) {
```
문득 네이밍이 부자연스럽다고 느껴지네요.
`PostService`의 경우엔 `showPosts`와 같이 `Post`를 명시하는데, 필터는 `filteredResults`로 하고 있군요.
`showFilteredPost`는 어떨까요?

YebinK replied:
동의합니다!
`PostFilterService.showFilteredResults`,
`PostFilterController.showFilteredResults` 모두 `showFilteredPosts`로 바꿔주었습니다!

기존에 showSearchResult에서 시작된 것이 아닐까 싶네요 ㅎㅎ

hwookim reviewed (approved):
querydsl은 참 어렵군요..... 왜 fetchjoin에는 on 할 수 없을까....
정말 고생 많으셨어요. 🤣

별달리 리뷰할 부분은 없는 것 같아요.
다만 전체적으로 다시 한번 코드 포맷팅을 돌려봐 주실 수 있을까요?
엘리가 바꾼 부분이 제 포맷터에서는 바뀌기 전으로 돌아오네요....🤔

yebink committed:
feature/177 refactor: 게시글 필터링 메소드명 리팩토링

 - PostFilterController.showFilteredResults -> showFilteredPosts
 - PostFilterSErvice.showFilteredResults -> showFilteredPosts

LTTTTTE commented on front/src/store/modules/post.js:
```diff
@@ -27,6 +27,7 @@ export default {
       likes: [],
       createdAt: "",
       comments: [],
+      entityStatus: "",
```
API요청하면 ALIVE인 친구들만 body에 넣어줄거같은데 response나 front에는 없어도 될거같은 생각이드네요🐰
혹시 필요했다면 어디서 필요했나요?🥕

YejiAhn commented on front/src/store/modules/post.js:
```diff
@@ -27,6 +27,7 @@ export default {
       likes: [],
       createdAt: "",
       comments: [],
+      entityStatus: "",
```
동의합니다! 프론트에서 어디에 쓰이나 찾아봤는데 쓰이는 곳도 없는 것 같네요 🙂 API 요청을 통해 살아있는/삭제된 인스턴스 전부 다 가져오지 않는 이상 이 필드가 쓰일 경우가 무엇이 있을까요?

hwookim replied:
오 이 문제는 생각을 못했었는데, 예지 리뷰 읽어보니 너무 합당한 거 같아요.
저도 해당 필드는 프론트에서 필요 없다는 생각이 드네요!

YejiAhn commented on src/main/java/com/grasshouse/dorandoran/comment/domain/Comment.java:
```diff
@@ -69,7 +76,7 @@ public Comment(Long id, Member author, Post post, String content, Double distanc
 
     private void setAuthor(Member author) {
-        if (Objects.isNull(author)) {
+        if (Objects.isNull(this.author)) {
```
👍

YejiAhn commented on src/main/java/com/grasshouse/dorandoran/common/baseentity/EntityStatus.java:
```diff
@@ -0,0 +1,8 @@
+
+    ALIVE,
+    DELETED,
+
```
불필요한 개행 아닐까요?

YejiAhn reviewed (approved):
간단한 코멘트 몇 개 남겼어요!
글과 댓글들을 어드민 입장에서 더 잘 관리할 수 있게 되었군요.
최고에요 엘리 💯
